File: src/leetcode/125.valid_palindrome/valid_palidrome.py
import unittest
from typing import *


class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = "".join(c for c in s.lower() if c.isalnum())
        left = 0
        right = len(s) - 1
        middle = right / 2

        while left < middle:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1

        return True

    def isPalindromeRevert(self, s: str) -> bool:
        s = "".join(c for c in s.lower() if c.isalnum())
        return s == s[::-1]


class TestValidPalidrome(unittest.TestCase):
    def test_sample_1(self):
        s = "A man, a plan, a canal: Panama"
        self.assertEqual(Solution().isPalindrome(s), True)

    def test_sample_2(self):
        s = "race a car"
        self.assertEqual(Solution().isPalindrome(s), False)

    def test_sample_3(self):
        s = "    "
        self.assertEqual(Solution().isPalindrome(s), True)

    def test_sample_4(self):
        s = "0P"
        self.assertEqual(Solution().isPalindrome(s), False)


if __name__ == "__main__":
    unittest.main()

File: src/leetcode/125.valid_palindrome/valid_palidrome.rs
struct Solution;
impl Solution {
    pub fn is_palindrome(s: String) -> bool {
        let cleaned_str: String = s
            .chars()
            .filter(|c| c.is_alphanumeric())
            .flat_map(|c| c.to_lowercase())
            .collect();
        println!("{}", cleaned_str);

        let length = cleaned_str.len();

        let byte_str = cleaned_str.as_bytes();

        for left in 0..length / 2 {
            if byte_str[left] != byte_str[length - 1 - left] {
                return false;
            }
        }
        true
    }
}

fn main() {
    let s = String::from("A man, a plan, a canal: Panama");
    let x = Solution::is_palindrome(s);
    println!("{}", x);
}

#[cfg(test)]
mod tests {
    use super::Solution;

    // Write your tests here
    #[test]
    fn test_case_1() {
        let s = String::from("A man, a plan, a canal: Panama");

        assert_eq!(Solution::is_palindrome(s), true);
    }
    #[test]
    fn test_case_2() {
        let s = String::from("race a car");

        assert_eq!(Solution::is_palindrome(s), false);
    }
    #[test]
    fn test_case_3() {
        let s = String::from(" ");

        assert_eq!(Solution::is_palindrome(s), true);
    }
}
